docs/RefMan/OverloadedOperations.rst

```diff
@@ -254,3 +280,71 @@ Rounding
     - ``ValidFloat e p``
 
 
+.. _derived-instances:
+
+Derived instances
+-----------------
+
+A :ref:`newtype <newtypes>` or :ref:`enum <enums>` declaration can optionally
+have a ``deriving`` clause attached to it to *derive* instances of the type for
+certain constraints. Like built-in instances, the derived instances may have
+certain conditions that need to be satisifed for them to be valid.
+
+Newtypes
+~~~~~~~~
+
+The following constraints can be derived for newtypes:
+
+  * ``Eq``
+  * ``Cmp``
+  * ``SignedCmp``
+  * ``Zero``
+  * ``Logic``
+  * ``Ring``
+
+The condition for a newtype to be an instance of any of the above constraints is
+that the underlying record type of the newtype is also an instance of that
+constraint. For instance, if you have
+
+.. code-block:: cryptol
+
+  newtype T a b = { f1 : a, f2 : b } deriving (Eq, Cmp)
+
+then ``Eq (T a b)`` will only hold if ``Eq ({f1 : a, f2 : b})`` holds, and ``Cmp
+(T a b)`` will only hold if ``Cmp ({f1 : a, f2 : b})`` holds.
+
+The behavior of the built-in overloaded operations on newtypes with derived
+instances is exactly the same as the behavior on their underlying record types,
+modulo the appropriate wrapping and unwrapping of the newtype constructor.
+
+Enums
+~~~~~
+
+The following constraints can be derived for enums:
+
+  * ``Eq``
+  * ``Cmp``
+  * ``SignedCmp``
+
+The condition for an enum to be an instance of any of the above constraints is
+that all the fields of all its constructors are instances of that constraint.
+For instance, if you have
+
+.. code-block:: cryptol
+
+  enum T a b = A a | B b deriving (Eq, Cmp)
+
+then ``Eq (T a b)`` will only hold if ``Eq a`` and ``Eq b`` hold, and ``Cmp (T a
+b)`` will only hold if ``Cmp a`` and ``Cmp b`` hold.
+
+Two enum values of the same type compare equal if they have the same constructor
+and the same field values for that constructor, and unequal otherwise.
```

Comment:
It might be useful to illustrate this point with some examples, such as the results of evaluating things like `A 0x0 == A 0x0`, `A 0x0 == B 0x0`, `A 0x0 == A 0x1`, etc.